src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -872,64 +923,111 @@ CreateDistributedExecution(RowModifyLevel modLevel, List *taskList, bool hasRetu
 
 
 /*
- * StartDistributedExecution sets up the coordinated transaction and 2PC for
- * the execution whenever necessary. It also keeps track of parallel relation
- * accesses to enforce restrictions that arise due to foreign keys to reference
- * tables.
+ * DecideTransactionProperties decides whether to initiate a coordinated transaction
+ * and whether to use 2PC for the given task list. These are global settings that
+ * affect the remainder of the transaction.
+ *
+ * Additionally, it returns whether the current execution should error out on any failure
+ * and whether to use transaction blocks, which may depend on previous statements in the
+ * transaction.
  */
-void
-StartDistributedExecution(DistributedExecution *execution)
+static void
+DecideTransactionProperties(RowModifyLevel modLevel, List *taskList,
+							bool *errorOnAnyFailure,
+							TransactionBlocksUsage *useTransactionBlocks)
 {
-	List *taskList = execution->tasksToExecute;
+	*errorOnAnyFailure = false;
+	*useTransactionBlocks = TRANSACTION_BLOCKS_ALLOWED;
 
-	if (MultiShardCommitProtocol != COMMIT_PROTOCOL_BARE)
+	if (taskList == NIL)
```

Comment:
From the coverage report it seems that taskList can be NIL, how can it be NIL? wouldnt it be better to return early in `AdaptiveExecutor` method so that we dont have to worry about this case in the rest of the method.

Reply:
> From the coverage report it seems that taskList can be NIL, how can it be NIL?

tip: Run the regression tests by adding an error message, you'd  see which tests fail so you'd have a better idea on what types of queries/commands are affected by it
```C
diff --git a/src/backend/distributed/executor/adaptive_executor.c b/src/backend/distributed/executor/adaptive_executor.c
index abb9d351b..81e90edd7 100644
--- a/src/backend/distributed/executor/adaptive_executor.c
+++ b/src/backend/distributed/executor/adaptive_executor.c
@@ -941,6 +941,7 @@ DecideTransactionProperties(RowModifyLevel modLevel, List *taskList,
 
        if (taskList == NIL)
        {
+               elog(ERROR, "which queries fail");
                /* nothing to do */
                return;
        }
```
>  wouldnt it be better to return early in AdaptiveExecutor method so that we dont have to worry about this case in the rest of the method.

There seem to be some side effects  like acquiring some locks `LockPartitionsForDistributedPlan()`, or  things in `FinishDistributedExecution()`.

It might be possible to early return in adaptive executor,  but we should be careful  about  those side-effects, and I would probably not do  it in this PR, maybe with a PR focusing on that.